### PR TITLE
Ck/11724 ts engine

### LINKED ISSUE #11724: [TS MVP] Rewrite ckeditor5-engine
Rewrite `ckeditor5-engine` to TypeScript.

Make sure that automated and manual tests run after the rewrite. Tests themselves are out of scope.

This would be the first serious application of `EmitterMixing` and typing of `Emitter` done in #11715.

### PR DESCRIPTION
**Pull request by arkflpc**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (engine): Rewrite `ckeditor5-engine` package to TypeScript. Closes #11724.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-engine/tests/view/utils-tests/createroot.js [hunk 1]
```diff
@@ -3,8 +3,8 @@
  * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
  */
 
-import Document from '../../../src/view/document.js';
-import RootAttributeElement from '../../../src/view/rooteditableelement.js';
+import Document from '../../../src/view/document.ts';
+import RootAttributeElement from '../../../src/view/rooteditableelement.ts';
 import createRoot from '../_utils/createroot.js';
```
Comment (niegowski) [inline packages/ckeditor5-engine/tests/view/utils-tests/createroot.js:8]: Those file extensions should not be there.

File: packages/ckeditor5-engine/src/controller/datacontroller.ts [hunk 2]
```diff
@@ -80,8 +105,8 @@ export default class DataController {
 			mapper: this.mapper,
 			schema: model.schema
 		} );
-		this.downcastDispatcher.on( 'insert:$text', insertText(), { priority: 'lowest' } );
-		this.downcastDispatcher.on( 'insert', insertAttributesAndChildren(), { priority: 'lowest' } );
+		this.downcastDispatcher.on<InsertEvent<ModelText | ModelTextProxy>>( 'insert:$text', insertText(), { priority: 'lowest' } );
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/controller/datacontroller.ts:108]: Maybe we should prefix those event types to indicate distinction between upcast and downcast events?

File: packages/ckeditor5-engine/src/view/treewalker.ts [hunk 3]
```diff
@@ -54,7 +56,7 @@ export default class TreeWalker {
 			 *
 			 * @error view-tree-walker-unknown-direction
 			 */
-			throw new CKEditorError( 'view-tree-walker-unknown-direction', options.startPosition, { direction: options.direction } );
+			throw new CKEditorError( 'view-tree-walker-unknown-direction', options, { direction: options.direction } );
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/view/treewalker.ts:59]: The `context` param is used to check which instance of the editor caused that exception. I'm not sure if the original implementation was valid but the new one might be broken? Is it possible to track the connection between that object and editor instance?

**Review by arkflpc — COMMENTED** [review]
Add typed events for view/observer/*

File: packages/ckeditor5-engine/src/view/document.ts [hunk 4]
```diff
@@ -218,8 +236,7 @@ export default class Document {
 	// @if CK_DEBUG_ENGINE // }
 }
 
-mix( Document, BubblingEmitterMixin );
-mix( Document, ObservableMixin );
+type ViewDocumentPostFixer = ( writer: DowncastWriter ) => boolean;
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/document.ts:239]: I'd export that

File: packages/ckeditor5-engine/src/view/documentfragment.ts [hunk 5]
```diff
@@ -98,37 +109,18 @@ export default class DocumentFragment {
...
-
 	/**
 	 * {@link module:engine/view/documentfragment~DocumentFragment#_insertChild Insert} a child node or a list of child nodes at the end
 	 * and sets the parent of these nodes to this fragment.
 	 *
 	 * @param {module:engine/view/item~Item|Iterable.<module:engine/view/item~Item>} items Items to be inserted.
 	 * @returns {Number} Number of appended nodes.
 	 */
-	_appendChild( items ) {
+	public _appendChild( items: Item | Iterable<Item> ): number {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/documentfragment.ts:123]: internal

File: packages/ckeditor5-engine/src/view/documentfragment.ts [hunk 6]
```diff
@@ -169,7 +161,7 @@ export default class DocumentFragment {
 	 * @param {module:engine/view/item~Item|Iterable.<module:engine/view/item~Item>} items Items to be inserted.
 	 * @returns {Number} Number of inserted nodes.
 	 */
-	_insertChild( index, items ) {
+	public _insertChild( index: number, items: Item | Iterable<Item> ): number {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/documentfragment.ts:164]: internal

File: packages/ckeditor5-engine/src/view/documentselection.ts [hunk 7]
```diff
@@ -180,16 +195,16 @@ export default class DocumentSelection {
 	 *
 	 * @protected
 	 */
-	get _ranges() {
-		return this._selection._ranges;
+	public get _ranges(): Range[] {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/documentselection.ts:198]: internal

File: packages/ckeditor5-engine/src/view/documentselection.ts [hunk 8]
```diff
@@ -357,8 +348,8 @@ export default class DocumentSelection {
 	 * @param {Boolean} [options.fake] Sets this selection instance to be marked as `fake`.
 	 * @param {String} [options.label] Label for the fake selection.
 	 */
-	_setTo( selectable, placeOrOffset, options ) {
-		this._selection.setTo( selectable, placeOrOffset, options );
+	public _setTo( ...args: Parameters< Selection[ 'setTo' ]> ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/documentselection.ts:351]: internal

File: packages/ckeditor5-engine/src/view/documentselection.ts [hunk 9]
```diff
@@ -373,7 +364,7 @@ export default class DocumentSelection {
 	 * @param {Number|'end'|'before'|'after'} [offset] Offset or one of the flags. Used only when
 	 * first parameter is a {@link module:engine/view/item~Item view item}.
 	 */
-	_setFocus( itemOrPosition, offset ) {
+	public _setFocus( itemOrPosition: Item | Position, offset?: number | 'before' | 'end' | 'after' ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/documentselection.ts:367]: internal

File: packages/ckeditor5-engine/src/view/downcastwriter.ts [hunk 10]
```diff
@@ -194,19 +208,27 @@ export default class DowncastWriter {
 	 * pipeline even though they would normally be filtered out by unsafe attribute detection mechanisms.
 	 * @returns {module:engine/view/attributeelement~AttributeElement} Created element.
 	 */
-	createAttributeElement( name, attributes, options = {} ) {
+	public createAttributeElement(
+		name: string,
+		attributes?: Record<string, string> | Iterable<[ string, string ]>,
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/downcastwriter.ts:213]: ConstructorParameters<typeof Element>[2] would be consistent

File: packages/ckeditor5-engine/src/view/downcastwriter.ts [hunk 11]
```diff
@@ -1259,7 +1328,7 @@ export default class DowncastWriter {
 	 * @protected
 	 * @param {Function} slotFactory The slot factory.
 	 */
-	_registerSlotFactory( slotFactory ) {
+	public _registerSlotFactory( slotFactory: ( writer: DowncastWriter, modeOrFilter: string | SlotFilter ) => Element ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/downcastwriter.ts:1331]: internal

File: packages/ckeditor5-engine/src/view/downcastwriter.ts [hunk 12]
```diff
@@ -1268,7 +1337,7 @@ export default class DowncastWriter {
 	 *
 	 * @protected
 	 */
-	_clearSlotFactory() {
+	public _clearSlotFactory(): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/downcastwriter.ts:1340]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 13]
```diff
@@ -583,7 +570,7 @@ export default class Element extends Node {
 	 * element will be cloned without any children.
 	 * @returns {module:engine/view/element~Element} Clone of this element.
 	 */
-	_clone( deep = false ) {
+	public _clone( deep = false ): Element {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:573]: internal
also check other _clone in this package

File: packages/ckeditor5-engine/src/view/element.ts [hunk 14]
```diff
@@ -624,22 +611,23 @@ export default class Element extends Node {
 	 * @fires module:engine/view/node~Node#change
 	 * @returns {Number} Number of appended nodes.
 	 */
-	_appendChild( items ) {
+	public _appendChild( items: Item | Iterable<Item> ): number {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:614]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 15]
```diff
@@ -624,22 +611,23 @@ export default class Element extends Node {
...
+	 * @internal
 	 * @see module:engine/view/downcastwriter~DowncastWriter#insert
 	 * @protected
 	 * @param {Number} index Position where nodes should be inserted.
 	 * @param {module:engine/view/item~Item|Iterable.<module:engine/view/item~Item>} items Items to be inserted.
 	 * @fires module:engine/view/node~Node#change
 	 * @returns {Number} Number of inserted nodes.
 	 */
-	_insertChild( index, items ) {
+	public _insertChild( index: number, items: Item | Iterable<Item> ): number {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:630]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 16]
```diff
@@ -672,7 +660,7 @@ export default class Element extends Node {
 	 * @fires module:engine/view/node~Node#change
 	 * @returns {Array.<module:engine/view/node~Node>} The array of removed nodes.
 	 */
-	_removeChildren( index, howMany = 1 ) {
+	public _removeChildren( index: number, howMany: number = 1 ): Node[] {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:663]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 17]
```diff
@@ -691,7 +679,7 @@ export default class Element extends Node {
 	 * @param {String} value Attribute value.
 	 * @fires module:engine/view/node~Node#change
 	 */
-	_setAttribute( key, value ) {
+	public _setAttribute( key: string, value: string ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:682]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 18]
```diff
@@ -714,7 +702,7 @@ export default class Element extends Node {
 	 * @returns {Boolean} Returns true if an attribute existed and has been removed.
 	 * @fires module:engine/view/node~Node#change
 	 */
-	_removeAttribute( key ) {
+	public _removeAttribute( key: string ): boolean {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:705]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 19]
```diff
@@ -754,7 +742,7 @@ export default class Element extends Node {
 	 * @param {Array.<String>|String} className
 	 * @fires module:engine/view/node~Node#change
 	 */
-	_addClass( className ) {
+	public _addClass( className: string | string[] ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:745]: internal

File: packages/ckeditor5-engine/src/view/element.ts [hunk 20]
```diff
@@ -773,7 +761,7 @@ export default class Element extends Node {
 	 * @param {Array.<String>|String} className
 	 * @fires module:engine/view/node~Node#change
 	 */
-	_removeClass( className ) {
+	public _removeClass( className: string | string[] ): void {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/element.ts:764]: internal

File: packages/ckeditor5-engine/src/view/upcastwriter.ts [hunk 21]
```diff
@@ -360,7 +373,7 @@ export default class UpcastWriter {
 	 * first parameter is a {@link module:engine/view/item~Item view item}.
 	 * @returns {module:engine/view/position~Position}
 	 */
-	createPositionAt( itemOrPosition, offset ) {
+	public createPositionAt( itemOrPosition: Item | Position, offset: number | 'before' | 'after' | 'end' ): Position {
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/upcastwriter.ts:376]: offset**?**

File: packages/ckeditor5-engine/src/view/rawelement.ts [hunk 22]
```diff
@@ -136,8 +103,53 @@ export default class RawElement extends Element {
 	 * @param {HTMLElement} domElement The native DOM element representing the raw view element.
 	 * @param {module:engine/view/domconverter~DomConverter} domConverter Instance of the DomConverter used to optimize the output.
 	 */
+	public render( domElement?: DomElement, domConverter?: DomConverter ): void;
+	public render(): void {}
+	// TODO
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/view/rawelement.ts:108]: I'm not sure about what is this TODO?

File: packages/ckeditor5-engine/src/view/observer/bubblingemittermixin.ts [hunk 23]
```diff
@@ -24,123 +36,144 @@ const contextsSymbol = Symbol( 'bubbling contexts' );
...
-			let emitter = eventContexts.get( context );
+			for ( const context of contexts ) {
+				let emitter = eventContexts.get( context );
 
-			if ( !emitter ) {
-				emitter = Object.create( EmitterMixin );
-				eventContexts.set( context, emitter );
-			}
+				if ( !emitter ) {
+					emitter = Object.create( EmitterMixin );
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/view/observer/bubblingemittermixin.ts:135]: Is this correct with new mixins?

**Review by pomek — APPROVED** [review]
I checked the release-oriented things and LGTM.

File: packages/ckeditor5-engine/src/model/position.ts [hunk 24]
```diff
@@ -544,24 +571,9 @@ export default class Position {
...
-	 *
-	 * {@link module:engine/model/node~Node#is Check the entire list of model objects} which implement the `is()` method.
-	 *
-	 * @param {String} type
-	 * @returns {Boolean}
-	 */
-	is( type ) {
-		return type === 'position' || type === 'model:position';
+		/* istanbul ignore next */
+		throw new Error( 'unreachable code' );
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/position.ts:576]: I'm not sure if this the best way. Let's keep some comment on whys.

**Pull request merged**